API specification for multi-asset wallet migration

jonathanknowles opened the pull request:
# Issue Number

ADP-840

# Overview

This PR revises the **_API specification_** to support multi-asset wallet migrations.

The **_server implementation_** will be added in a subsequent PR. (Connecting the API with the algorithm.)

# Details

This PR:

- [x] Adds a `ApiWalletMigrationBalance` type:

    ```yaml
    ApiWalletMigrationBalance: &ApiWalletMigrationBalance
      type: object
      required:
        - ada
        - assets
      properties:
        ada: *amount
        assets: *walletAssets
    ```

- [x] Revises the `ApiWalletMigrationInfo` type:

    ```yaml
    ApiWalletMigrationInfo: &ApiWalletMigrationInfo
      type: object
      required:
        - selections
        - total_fee
        - balance_leftover
        - balance_selected
      properties:
        selections:
          type: array
          items: *ApiCoinSelection
        total_fee:
          <<: *amount
        balance_leftover:
          <<: *ApiWalletMigrationBalance
        balance_selected:
          <<: *ApiWalletMigrationBalance

    ```

piotr-iohk commented on lib/core/src/Cardano/Wallet/Api/Types.hs:
```diff
@@ -993,6 +995,18 @@ data ApiPostRandomAddressData = ApiPostRandomAddressData
+
 data ApiWalletMigrationPostData (n :: NetworkDiscriminant) (s :: Symbol) =
     ApiWalletMigrationPostData
     { passphrase :: !(ApiT (Passphrase s))
```
passphrase wouldn't be needed here, but in the next step (i.e. confirming migration) if I'm not mistaken?
![Screenshot from 2021-03-31 13-21-24](https://user-images.githubusercontent.com/42900201/113136771-0973e400-9224-11eb-959a-ff1dd0ccf7a9.png)

Anviking replied:
Also, is POST the correct verb now when the selection is only returned and not submitted?

I guess that depends on whether we mark the tx as pending immediately or not, which was a point of uncertainty in https://jira.iohk.io/browse/ADP-342

jonathanknowles replied:
> I guess that depends on whether we mark the tx as pending immediately or not.

The intention is for the user to:

1. Call the `migrateWallet` endpoint, which supplies a list of selections (= unsigned transactions).
2. Inspect the results, check that the total fee and the list of selections are acceptable.
3. Call another endpoint (yet-to-be-added), which accepts a list of selections. In the case of HW wallets, the user would sign them first.

The yet-to-be-added endpoint would check that all the supplied selections are valid, by checking that all the UTxO entries still exist, before submitting them as transactions to the network.

So I think there's no need for the `migrateWallet` endpoint to mark transactions as pending here.

If the user calls the `migrateWallet` endpoint and then makes another transaction that spends some of the UTxO entries in the meantime, they can easily get themselves unstuck by calling `migrateWallet` again to generate a new set of selections that don't include those UTxO entries.

jonathanknowles replied:
> Also, is POST the correct verb now when the selection is only returned and not submitted?

I suspect POST is still the correct verb, as the user still has to post a list of target addresses so that we can generate real coin selections with `(Address, TokenBundle)` pairs as the outputs.

jonathanknowles replied:
> passphrase wouldn't be needed here, but in the next step (i.e. confirming migration) if I'm not mistaken?

Ah yes, well-spotted!

You're right: only the final step (calling the yet-to-be-added submission endpoint) should require a passphrase.

jonathanknowles replied:
> passphrase wouldn't be needed here, but in the next step (i.e. confirming migration) if I'm not mistaken?

Fixed in ed63066f2360baaaef90fb15337e636207885e74.

jonathanknowles replied:
@piotr-iohk @Anviking 

This has changed: we now keep the existing API endpoints (without changing them), but just modify the `ApiWalletMigrationInfo` type. (See updated PR description.)

Anviking commented on lib/core/src/Cardano/Wallet/Api/Types.hs:
```diff
@@ -993,6 +995,18 @@ data ApiPostRandomAddressData = ApiPostRandomAddressData
 
+data ApiWalletMigration (n :: NetworkDiscriminant) = ApiWalletMigration
+    { selections :: !(NonEmpty (ApiCoinSelection n))
+    , cost :: !(Quantity "lovelace" Natural)
```
No strong opinion, but if this was named `fee` it would be consistent with `ApiTransaction`.

jonathanknowles replied:
> No strong opinion, but if this was named fee it would be consistent with ApiTransaction.

Good point! Perhaps `totalFee` would be even better, because each individual selection has its own fee, and this is the sum of all fees. What do you think?

jonathanknowles replied:
Changed this to `totalFee` in commit 03696719acfd86b225d6fca8f11766f911a0b1c0.

Anviking commented on specifications/api/swagger.yaml:
```diff
@@ -1628,6 +1628,30 @@ components:
+          items: *ApiCoinSelection
+        cost: *amount
+        balance_included: *ApiWalletMigrationBalance
+        balance_leftover: *ApiWalletMigrationBalance
```
I guess we should explain why there might be leftovers, now or later, such that users know if they need to care.

Assets could be leftover in some cases, I recall. Could dust ada end up here too?

Edit: Looking at the previous description:

> Leftovers dust coins which won't be migrated nor spent as fees.

I'd guess so too? 

And I see you're on it:

> TODO: Updates the API endpoint documentation.

jonathanknowles replied:
> I guess we should explain why there might be leftovers, now or later, such that users know if they need to care.

Definitely!

> Assets could be leftover in some cases, I recall.

That's right. The requirements allow us to omit assets from a migration if the algorithm cannot find a viable strategy to include them. An example of when this might happen is if the remaining UTxO contains a single entry with the minimum ada amount and a single non-ada asset. In this situation, it's impossible to create a transaction, because there's not enough ada to pay for the fee (the only source of ada is the minimum ada amount, which can't be reduced). The only other option is to burn the assets, which the requirements do not allow.

> Could dust ada end up here too?

Ada quantities that are less than or equal to `input_marginal_fee` will be omitted by the algorithm automatically, and not even considered for inclusion.

It is also possible for ada quantities greater than `input_marginal_fee` to be omitted: if the remaining UTxO set contains only **_n_** small values whose total value is less than (`minimum_ada_quantity` + `base_transaction_fee` + **_n_** × `input_marginal_fee`), then it won't be possible to create a transaction from them.

Anviking reviewed (commented):
Looks reasonable, I think

jonathanknowles committed:
[Reformatting] Tidy up migration part of `Api.TypesSpec`.

jonathanknowles committed:
Add `ApiWalletMigrationBalance` type.

jonathanknowles committed:
Rename field `leftovers` to `balanceLeftover`.

We'll eventually need the `ApiWalletMigrationInfo` record to encode two
different types of balance:

- balanceSelected
- balanceLeftover

jonathanknowles committed:
Change type of `balanceLeftover` to `ApiWalletMigrationBalance`.

jonathanknowles committed:
Add `balanceSelected` field to `ApiWalletMigrationInfo`.

jonathanknowles committed:
Rename `migration_cost` to `total_fee`.

jonathanknowles committed:
Add `selections` field to `ApiWalletMigrationInfo`.

jonathanknowles committed:
Add documentation to specification for `ApiWalletMigrationInfo`.

KtorZ commented on lib/core/test/unit/Cardano/Wallet/Api/TypesSpec.hs:
```diff
@@ -1279,7 +1295,13 @@ instance Arbitrary ApiByronWalletBalance where
+    shrink = genericShrink
+
+instance Arbitrary (ApiWalletMigrationInfo n) where
     arbitrary = genericArbitrary
```
![Screenshot from 2021-04-29 07-25-30](https://user-images.githubusercontent.com/5680256/116505834-31567600-a8bc-11eb-8a4a-34a0b41d5c86.png)

Be careful with genericArbitrary! This generates data-set that are way too large!

jonathanknowles replied:
> Be careful with genericArbitrary! This generates data-set that are way too large!

Thanks for spotting this! I'll revise the `Arbitrary` instance to generate a smaller data set. 👍🏻

rvl commented on specifications/api/swagger.yaml:
```diff
@@ -2097,18 +2097,50 @@ components:
-          <<: *amount
-          description: Total amount which will be paid as fees for the migration.
-        leftovers:
+        selections:
```
This field could have a description too.

jonathanknowles replied:
> This field could have a description too.

That's a good point!

I've added a description in this updated commit: 202e0240bfbfcdb88bbdb3ef780dce36a13cb890.

rvl reviewed (commented):
There should also be updates to the `getShelleyWalletMigrationInfo` and `migrateShelleyWallet` specs, right?

jonathanknowles committed:
Revise `migrate{Byron,Shelley}Wallet` endpoint descriptions.

jonathanknowles committed:
Revise `get{Byron,Shelley}WalletMigrationInfo` endpoint descriptions.

rvl reviewed (approved):
Nice thanks - better now.
